README.md
# cloudwatch-logs-search

This is sample logic that filters and retrieves logs from Amazon CloudWatch.

## Reference
- https://wp-kyoto.net/get-cloudwatchlogs-data-by-javascript/
- https://docs.aws.amazon.com/ja_jp/awscloudtrail/latest/userguide/awscloudtrail-ug.pdf
- https://business.ntt-east.co.jp/content/cloudsolution/column-try-28.html#section-04

## Changelog

See [CHANGELOG.md](./CHANGELOG.md).

## Getting Started

Install package.  

```sh
npm i cloudwatch-logs-search;
```

Install dependent packages.  

```sh
cd cloudwatch-logs-search;
npm i;
```

Add your Amazon CloudWatch Logs credentials to your environment variables.  

.env:  

```text
NODE_ENV=development
AWS_CLOUD_WATCH_LOGS_REGION=ap-northeast-1
AWS_CLOUD_WATCH_LOGS_ACCESS_KEY=****
AWS_CLOUD_WATCH_LOGS_SECRET_KEY=********
```

Search logs.  

The log group name is an exact match search and the log stream name is an fuzzy search.  

```sh
node search.js 'Log group name' 'Log stream name';
```

## Usage

How to use search.js.  

Gets all logs from the stream that contains the specified log stream name..  

```sh
node search.js 'Log group name' 'Log stream name';
```

Gets the log that matches the keyword from the stream that contains the specified log stream name..  

```sh
node search.js 'Log group name' 'Log stream name' 'keyword';
```

## License

[MIT licensed](./LICENSE.txt)

CHANGELOG.md
# Changelog

### [1.0.6] - 2021-4-16

* Refactor JS.

### [1.0.5] - 2021-4-12

* Cloudwatch credentials changed to set in constructor.

### [1.0.4] - 2021-4-12

* Fixed to get all log streams recursively.

shared/SNS.js
const AWS = require('aws-sdk');

/**
 * SNS notification client.
 */
module.exports = new class {
  /**
   * Generate a SNS client instance.
   */
  constructor() {
    this.client = new AWS.SNS({apiVersion: '2010-03-31'});
  }

  /**
   * Send a message to SNS.
   */
  async postMessage(message) {
  // async postMessage(subject, message) {
    return new Promise((resolve, reject) => {
      this.client.publish({
        // Subject: subject,
        Message: message,
        TopicArn: process.env.SNS_TOPIC_ARN
      })
        .promise()
        .then(data => {
          console.log(`Message "${message.length > 30 ? message.slice(0, 30) + '...' : message}" sent to the topic ${process.env.SNS_TOPIC_ARN}`);
          console.log(`MessageID is ${data.MessageId}`);
          resolve(data);
        })
        .catch(err => reject(err));
    });
  }
}

src/findCloudWatchLogs.js
const AWS = require('aws-sdk')
const moment = require('moment')

/**
 * Find logs from cloudwatch.
 * 
 * @param  {string} options.logGroupName: Log group name you want to find.
 * @param  {string} options.keyword: Keywords for filtering logs.
 * @param  {number} options.startTime: Search period start date. UnixTime milliseconds.The default is yesterday's 00:00:00.
 * @param  {number} options.endTime: Search period end date. UnixTime milliseconds.The default is yesterday at 23:59:59.
 */
module.exports = async (options = {}) => {
  try {
    // Initialize options.
    options = Object.assign({
      logGroupName: undefined,
      logStreamName: undefined,
      keyword: undefined,
      startTime: moment().add(-1, 'days').startOf('day').valueOf(),
      endTime: moment().add(-1, 'days').endOf('day').valueOf(),
      debug: false
    }, options);

    // assign AWS credentials here in following way.
    const result = require('dotenv').config({path: '.env'})
    if (result.error) throw result.error
    const credentials = result.parsed;
    AWS.config.update({
      accessKeyId: credentials.AWS_CLOUD_WATCH_LOGS_ACCESS_KEY,
      secretAccessKey: credentials.AWS_CLOUD_WATCH_LOGS_SECRET_KEY,
      region: credentials.AWS_CLOUD_WATCH_LOGS_REGION
    });

    // CloudWatch Logs instance.
    const cloudwatchlogs = new AWS.CloudWatchLogs({apiVersion: '2014-03-28'});

    // Get LogStreams.
    let streams = await (new Promise((resolve, reject) => {
      cloudwatchlogs.describeLogStreams({logGroupName: options.logGroupName}, (err, data) => {
        // an error occurred
        if (err) return void reject(err);
        // successful response
        resolve(data.logStreams);
      });
    }));

    // Get the start and end times of the previous day.
    console.log(`Start: ${moment(options.startTime).format('YYYY-MM-DD HH:mm:ss')}`);
    console.log(`End: ${moment(options.endTime).format('YYYY-MM-DD HH:mm:ss')}`);

    // Find the Nginx access log from a day ago.
    streams = streams.filter(stream => {
      // Filter by log stream name.
      if (options.logStreamName && stream.logStreamName.indexOf(options.logStreamName) === -1)
        return false;

      // Filtering by period.
      return !(stream.firstEventTimestamp < options.startTime && stream.lastEventTimestamp < options.startTime);
    });

    // End if there is no log of the previous day.
    if (!streams.length) return void console.log('There is no access log of the previous day.');

    // Debug found log stream.
    if (options.debug) {
      for (let {logStreamName, firstEventTimestamp, lastEventTimestamp} of streams) {
        console.log('==================================================');
        console.log(`logStreamName: ${logStreamName}`);
        console.log(`firstEventTimestamp: ${moment(firstEventTimestamp).format('YYYY-MM-DD HH:mm:ss')}`);
        console.log(`lastEventTimestamp: ${moment(lastEventTimestamp).format('YYYY-MM-DD HH:mm:ss')}`);
        console.log('==================================================');
      }
    }

    // Filter logs by keyword.
    let events = await (new Promise((resolve, reject) => {
      cloudwatchlogs.filterLogEvents({
       logGroupName: options.logGroupName,
       logStreamNames: streams.map(stream => stream.logStreamName),
       filterPattern: options.keyword ? `"${options.keyword}"` : undefined,
       limit: 10000,
       startTime: options.startTime,
       endTime: options.endTime
      }, (err, data) => {
        // an error occurred
        if (err) return void reject(err);
        // successful response
        resolve(data.events);
      });
    }));

    // // Debug found events.
    // for (let {timestamp, message} of events)
    //   console.log(`${moment(timestamp).format('YYYY-MM-DD HH:mm:ss')} -> message: ${message}`);

    // Return event log.
    return events.map(event => event.message);
  } catch (e) {
    console.error(e);
  }
}

shared/Cloudwatch.js
const AWS = require('aws-sdk');
const moment = require('./moment');

/**
 * Cloudwatch class.
 */
module.exports = new class {
  /**
   * Generate cloudWatch Logs instance.
   */
  constructor() {
    // assign AWS credentials here in following way.
    const result = require('dotenv').config({path: '.env'})
    if (result.error) throw result.error;

    // CloudWatch Logs instance.
    const credentials = result.parsed;
    this.client = new AWS.CloudWatchLogs({
      apiVersion: '2014-03-28',
      accessKeyId: credentials.AWS_CLOUD_WATCH_LOGS_ACCESS_KEY,
      secretAccessKey: credentials.AWS_CLOUD_WATCH_LOGS_SECRET_KEY,
      region: credentials.AWS_CLOUD_WATCH_LOGS_REGION
    });
  }

  /**
   * Find logs from cloudwatch.
   * 
   * @param  {string} options.logGroupName: Log group name you want to find.
   * @param  {string} options.keyword: Keywords for filtering logs.
   * @param  {number} options.since: Search period start date. UnixTime milliseconds.The default is yesterday's 00:00:00.
   * @param  {number} options.until: Search period end date. UnixTime milliseconds.The default is yesterday at 23:59:59.
   */
  async find(options) {
    try {
       // Initialize options.
      options = Object.assign({
        logGroupName: undefined,
        logStreamName: undefined,
        keyword: undefined,
        since: moment().add(-1, 'days').startOf('day').valueOf(),
        until: moment().add(-1, 'days').endOf('day').valueOf(),
        debug: false
      }, options);

      // Debug start and end dates.
      if (options.debug) {
        console.log(`Start: ${moment(options.since).format('YYYY-MM-DD HH:mm:ss')}`);
        console.log(`End: ${moment(options.until).format('YYYY-MM-DD HH:mm:ss')}`);
      }

      // Get LogStreams.
      let streams = await this.findLogStreams(options.logGroupName);

      // Find the Nginx access log from a day ago.
      streams = streams.filter(stream => {
        // Filter by log stream name.
        if (options.logStreamName && stream.logStreamName.indexOf(options.logStreamName) === -1)
          return false;

        // Filtering by period.
        return !(stream.firstEventTimestamp < options.since && stream.lastEventTimestamp < options.since);
      });

      // End if there is no log of the previous day.
      if (!streams.length) return void console.log('There is no access log of the previous day.');

      // Debug found log stream.
      if (options.debug) {
        for (let {logStreamName, firstEventTimestamp, lastEventTimestamp} of streams) {
          console.log('==================================================');
          console.log(`logStreamName: ${logStreamName}`);
          console.log(`firstEventTimestamp: ${moment(firstEventTimestamp).format('YYYY-MM-DD HH:mm:ss')}`);
          console.log(`lastEventTimestamp: ${moment(lastEventTimestamp).format('YYYY-MM-DD HH:mm:ss')}`);
          console.log('==================================================');
        }
      }

      // Filter logs by keyword.
      const events = await (new Promise((resolve, reject) => {
        this.client.filterLogEvents({
          logGroupName: options.logGroupName,
          logStreamNames: streams.map(stream => stream.logStreamName),
          filterPattern: options.keyword ? `"${options.keyword}"` : undefined,
          interleaved: true,
          limit: 10000,
          startTime: options.since,
          endTime: options.until
        }, (err, data) => {
          // an error occurred
          if (err) return void reject(err);
          // successful response
          resolve(data.events);
        });
      }));

      // Debug found events.
      for (let {timestamp, message} of events)
        console.log(`${moment(timestamp).format('YYYY-MM-DD HH:mm:ss')} -> message: ${message}`);

      // Return event log.
      return events.map(event => event.message);
    } catch (e) {
      console.error('Error searching cloudwatch logs: ', e);
      throw e;
    }
  }

  /**
   * Find Log Streams.
   */
  async findLogStreams(logGroupName) {
    return new Promise((resolve, reject) => {
      let streams = [];
      const recursive = (nextToken = undefined) => {
        this.client.describeLogStreams({logGroupName, nextToken}).promise()
          // successful response
          .then(data => {
            streams.push(...data.logStreams);
            if (data.nextToken) recursive(data.nextToken);
            else resolve(streams);
          })
          // an error occurred
          .catch(reject);
      };
      recursive();
    });
  }
}